### DISCUSSION TITLE
Refreshing tokens

### DISCUSSION THREAD

**Discussion by diogovieira** [microsoftgraph/msgraph-sdk-go#108] [Q&A] [answered]
Hello!

I’m trying to write a CLI tool that makes use of this SDK to query for events in the outlook calendar. So far I’m able to access the user’s events in the first run after interactive authentication with the browser or device code, using the provided examples in the documentation.

However, it is not clear how I can re-use and refresh tokens if necessary upon second run? Please keep in mind this is a tool that will make a query and exit, clearing memory in the process. So the second time this is run it’ll need to re-authenticate which is far from ideal.

Thank you in advance!

**Comment by baywet** [comment, marked as answer]
Hi @diogovieira,
Thanks for using the Go SDK and for reaching out.
Azure Identity, the library we're using under the covers to handle authentication, provides a [TokenCachePersistenceOptions class/struct](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/src/TokenCachePersistenceOptions.cs#L57) that you can assign to the credential (device code in your case) to get the library to handle token storage and retrieval in a secure way for you depending on the target OS. [see this example in dotnet for the CLI we're building](https://github.com/microsoftgraph/msgraph-cli-core/blob/dc6cefcce50f695bf6ac4215b983fd8b6e05ae92/src/Microsoft.Graph.Cli.Core/Authentication/AuthenticationServiceFactory.cs#L52)

Unfortunately the Go version of the library [does seem to support it yet](https://github.com/Azure/azure-sdk-for-go/search?q=TokenCachePersistenceOptions) (it's more recent). I'm going to suggest you reach out directly to them to inquire about the plans for the library to support it.
The alternative would be to store the refresh token yourself, but we usually don't recommend it as it's an easy way to compromise security when not done properly, and doing it properly is significant work.

**Reply by dadams39** [reply]
Has this capability been updated?

**Reply by baywet** [reply]
> Has this capability been updated?

Not as far as I can tell, dont hesitate to create issues over there to get answers from that team
